dali/python/nvidia/dali/pipeline.py

```diff
@@ -1455,7 +1455,82 @@ def _discriminate_args(func, **func_kwargs):
     return ctor_args, fn_args
 
 
-def pipeline_def(fn=None, **pipeline_kwargs):
+def _regroup_args(func, pipeline_def_kwargs, fn_call_kwargs):
+    """Regroup arguments that are directed into Pipeline object construction and those that
+    are passed into pipeline definition function.
+
+    Parameters
+    ----------
+    func : Callable
+        The pipeline definition function that is decorated.
+    pipeline_def_kwargs : Dict
+        Kwargs passed to the @pipeline_def
+    fn_call_kwargs : Dict
+        Kwargs passed when invoking the decorated function
+
+    Returns
+    -------
+    (Dict, Dict)
+        Pipeline kwargs, Function kwargs
+    """
+    ctor_args, fn_kwargs = _discriminate_args(func, **fn_call_kwargs)
+    pipeline_kwargs = {**pipeline_def_kwargs, **ctor_args}  # Merge and overwrite dict
+    return pipeline_kwargs, fn_kwargs
+
+
+def _preprocess_pipe_func(func, conditionals_on):
+    """Transform the pipeline defintion function if the conditionals are enabled
+    """
+    if conditionals_on:
+        return _conditionals._autograph.to_graph(func)
+    else:
+        return func
+
+
+def _preprocess_pipe_object(pipe, conditionals_on, args, fn_kwargs):
+    """Based on the conditional mode status, preprocess the pipeline object before the graph
+    is created.
+    """
+    if conditionals_on:
+        Pipeline.push_current(pipe)
+        pipe._conditionals_enabled = True
+        pipe._condition_stack = _conditionals._ConditionStack()
+        # Add all parameters to the pipeline as "know" nodes in the top scope.
+        for arg in args:
+            if isinstance(arg, DataNode):
+                _conditionals.register_data_nodes(arg)
+        for _, arg in fn_kwargs.items():
+            if isinstance(arg, DataNode):
+                _conditionals.register_data_nodes(arg)
+        Pipeline.pop_current()
```

Comment:
Shouldn't that be a context manager?

Reply:
Context manager doesn't work with debug mode, that's why I push and pop by hand.

Reply:
Oh. In what way it does not work? Is this because the `pipe` is somehow spetial in that case? How about using `contexlib` or just `try: finially:` to somehow make sure we won't accidentaly leak pipelines.

Reply:
It is explicitly disabled in debug mode (it raises an error), so people don't define pipelines via `with pipe` for debug mode - we need a function that we capture and are able to run.
I can add the `try: finally:` or similar to always release it.

Reply:
Done